docs/best-practices/minimize_optimize_joins.md

```diff
@@ -0,0 +1,64 @@
+- Only a limited subset of the columns will be queried, i.e. certain columns can be excluded from denormalization.
+- You have the capability to shift processing out of ClickHouse into upstream systems like Flink, where real-time enrichment or flattening can be managed.
+
+Not all data needs to be denormalized - focus on the attributes that are frequently queried. Also consider [materialized views](/best-practices/use-materialized-views) to incrementally compute aggregates instead of duplicating entire sub-tables. When schema updates are rare and latency is critical, denormalization offers the best performance trade-off.
+
+For a full guide on denormalizing data in ClickHouse see [here](/data-modeling/denormalization).
+
+## When JOINs are required {#when-joins-are-required}
+
+When JOINs are required, ensure you’re using **at least 24.12 and preferably the latest**, as JOIN performance continues to improve. As of ClickHouse 24.12, the query planner now automatically places the smaller table on the right side of the join for optimal performance - a task that previously had to be done manually. Even more enhancements are coming soon, including more aggressive filter pushdown and automatic re-ordering of multiple joins.
+
+Follow these best practices to improve JOIN performance:
+
+* **Avoid cartesian products** If a value on the left-hand side matches multiple values on the right-hand side, the JOIN will return multiple rows - the so-called cartesian product. If your use case doesn't need all matches from the right-hand side but just any single match, you can use `ANY` JOINs (e.g. `LEFT ANY JOIN`). They are faster and use less memory than regular JOINs.
+* **Reduce the sizes of JOINed tables** The runtime and memory consumption of JOINs grows proportionally with the sizes of the left and right tables. To reduce the amount of processed data by the JOIN, add additional filter conditions in the `WHERE` or `JOIN ON` clauses of the query. ClickHouse pushes filter conditions as deep as possible down in the query plan, usually before JOINs. If the filters are not pushed down automatically (for any reason), rewrite one side of the JOIN as a sub-query to force pushdown.
+* **Use direct JOINs via dictionaries if appropriate** Standard JOINs in ClickHouse are executed in two phases: a build phase which iterates the right-hand side to build a hash table, followed by a probe phase which iterates the left-hand side to find matching join partners via hash table lookups. If the right-hand side is a [dictionary](/dictionary) or another table engine with key-value characteristics (e.g. [EmbeddedRocksDB](/engines/table-engines/integrations/embedded-rocksdb) or the [Join table engine](/engines/table-engines/special/join)), then ClickHouse can use the "direct" join algorithm, which effectively removes the need to build a hash table, speeding up query processing. This works for `INNER` and `LEFT OUTER` JOINs and is preferred for real-time analytical workloads.
```

Comment:
Maybe mention that a dictionary by definition doesn't have duplicate keys and during data loading keys are silently deduplicated (last loaded key wins) - therefore for many-to-one relations if one loads the many side into a dictionary  (e.g. roles vs actors) that would result in silent data loss w.r.t the join result, many-to-mant relations not possible etc.